subprojects/launcher/src/main/java/org/gradle/launcher/daemon/configuration/BuildProcess.java

```diff
@@ -43,9 +43,19 @@ public boolean configureForBuild(DaemonParameters requiredBuildParameters) {
 
         final JvmOptions jvmOptions = new JvmOptions(new IdentityFileResolver());
         jvmOptions.systemProperties(getJvmOptions().getImmutableSystemProperties());
+        if (requiredBuildParameters.hasUserConfiguredImmutableJvmArgs()) {
+            jvmOptions.setJvmArgs(getJvmOptions().getJvmArgs());
+            jvmOptions.setMinHeapSize(getJvmOptions().getMinHeapSize());
+            jvmOptions.setMaxHeapSize(getJvmOptions().getMaxHeapSize());
+        }
+
         List<String> currentImmutables = jvmOptions.getAllImmutableJvmArgs();
         List<String> requiredImmutables = requiredBuildParameters.getEffectiveSingleUseJvmArgs();
-        requiredImmutables.removeAll(DaemonParameters.DEFAULT_JVM_ARGS);
+
+        if (!requiredBuildParameters.hasUserConfiguredImmutableJvmArgs()) {
```

Comment:
The case that would break in addition is the one where the user defined arguments correspond to the defaults.

I will add a test case for that.